File: Data_structures/Using_perl/Ch-01/BinarySearch.pl
#!/usr/bin/env perl

use 5.014;
use strict;
use warnings;

my @arrs = qw/ 1 3 5 7 9 /;
my $idx;

my $sizeof = @arrs;

$idx = BSearch(\@arrs, $sizeof, 7);
$idx == -1 ? say "탐색 실패" : say "타겟 저장 인덱스: $idx";

$idx = BSearch(\@arrs, $sizeof, 4);
$idx == -1 ? say "탐색 실패" : say "타겟 저장 인덱스: $idx";

sub BSearch {
    my ( $arrs, $len, $target ) = @_;
    my $first = 0;
    my $last  = $len - 1;
    my $mid;

    while ( $first <= $last )
    {
        $mid = int ( ($first + $last) / 2 );

        if ($target == $arrs->[$mid])
        {
            return $mid;
        }
        else {
            if ($target < $arrs->[$mid])
            {
                $last = $mid - 1;
            }
            else
            {
                $first = $mid + 1;
            }
        }
    }

    return -1;
}

File: Data_structures/Using_C/Ch-01/BinarySearch.c
#include <stdio.h>

int
BSearch (int  ar[],
         int  len,
         int target)
{
  int first = 0;
  int last = len - 1 ;
  int mid;

  while (first <= last)
    {
      mid = (first + last) / 2;

      if (target == ar[mid])
        {
          return mid;
        }
      else
        {
          if (target < ar[mid])
            last = mid -1;
          else
            first = mid + 1;
        }
    }

  return -1;
}

int
main (int   argc,
      char *argv[])
{
  int arr[] = {1, 3, 5, 7, 9};
  int idx;

  idx = BSearch (arr, sizeof(arr)/sizeof(int), 7);
  if (idx == -1)
    {
      printf ("탐색 실패 \n");
    }
  else
    {
      printf ("타겟 저장 인덱스: %d \n", idx);
    }

  idx = BSearch (arr, sizeof(arr) / sizeof (int), 4);
  if(idx == -1)
    {
      printf ("탐색 실패 \n");
    }
  else
    {
      printf ("타겟 저장 인덱스: %d \n", idx);
    }

  return 0;
}
